processor/src/main/java/org/mapstruct/ap/internal/model/source/selector/TypeSelector.java

```diff
@@ -81,11 +86,15 @@ public TypeSelector(TypeFactory typeFactory) {
         return result;
     }
 
-    private List<ParameterBinding> getAvailableParameterBindingsFromMethod(Method method, Type targetType) {
-        List<ParameterBinding> availableParams = new ArrayList<ParameterBinding>( method.getParameters().size() + 2 );
+    private List<ParameterBinding> getAvailableParameterBindingsFromMethod(Method method, Type targetType,
+        Assignment assignment) {
+        List<ParameterBinding> availableParams = new ArrayList<ParameterBinding>( method.getParameters().size() + 3 );
 
         availableParams.addAll( ParameterBinding.fromParameters( method.getParameters() ) );
         addMappingTargetAndTargetTypeBindings( availableParams, targetType );
+        if ( assignment != null ) {
+            availableParams.add( ParameterBinding.fromAssignment( assignment ) );
```

Comment:
I think if athe assignment is considered, then we need to ignore the source parameters of the method (we should still include ˋ@Contextˋ params, though)...